src/lib/core/CHIPTLVText.cpp

```diff
@@ -0,0 +1,231 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2015-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <core/CHIPTLVText.hpp>
+#include <core/CHIPTLVUtilities.hpp>
+#include <support/CodeUtils.h>
+#include <core/CHIPTLVTypes.h>
+#include <cstdint>
+
+namespace chip {
+
+namespace TLV {
+
+namespace Debug {
+
+static CHIP_ERROR Print(TLV::TLVReader &reader, PrinterFunc printerFunc, int tabLevel, bool isFirst);
+
+CHIP_ERROR Print(TLV::TLVReader &reader, PrinterFunc printerFunc) {
+    return Print(reader, printerFunc, 0, true);
+}
+
+void PrintTabs(PrinterFunc printerFunc, int tabLevel) 
+{
+    for (int i = 0; i < tabLevel; i++) {
+        printerFunc("\t");
+    }
+}
+
+void PrintTag(PrinterFunc printerFunc, uint64_t tag, int tabLevel, bool isFirst)
+{
+    if (!isFirst) {
+        printerFunc(",\n");
+    }
+
+    PrintTabs(printerFunc, tabLevel);
+    
+    if (IsContextTag(tag)) {
+        printerFunc("%" PRIu32 " = ", TagNumFromTag(tag));
+    }
+    else if (IsProfileTag(tag)) {
+        printerFunc("0x%" PRIx16 "::0x%" PRIx16 ":0x%" PRIx16 " = ", VendorIdFromTag(tag), ProfileNumFromTag(tag), TagNumFromTag(tag));
+    }
+}
+
+#pragma GCC diagnostic ignored "-Wstack-usage="
+
+void DefaultPrinter(const char *aFormat, ...) 
+{
+    va_list args;
+
+    va_start(args, aFormat);
+    vprintf(aFormat, args);
+    va_end(args);
+}
+
+CHIP_ERROR Print(TLV::TLVReader &reader, PrinterFunc printerFunc, int tabLevel, bool isFirst)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (tabLevel == 0) {
+        err = reader.Next();
+        SuccessOrExit(err);
+    }
+
+    PrintTag(printerFunc, reader.GetTag(), tabLevel, isFirst);
+    
+    switch (reader.GetType()) {
+        case kTLVType_Structure:
+        case kTLVType_Array:
+        case kTLVType_List:
+        {
+            TLVType containerType;
+            const char *openingBrace, *closingBrace;
+
+            if (reader.GetType() == kTLVType_Structure) {
+                openingBrace = "{";
+                closingBrace = "}";
+            }
+            else if (reader.GetType() == kTLVType_Array) {
+                openingBrace = "[";
+                closingBrace = "]";
+            }
+            else {
+                openingBrace = "[[";
+                closingBrace = "]]";
+            }
+
+            if (IsContextTag(reader.GetTag()) || IsProfileTag(reader.GetTag())) {
```

Comment:
This seems tightly coupled to whether PrintTag printed the tag.  Could PrintTag return this boolean instead?